### PR TITLE
[Enhancement][BugFix] simplify cancel process of spilling

### PR DESCRIPTION
**Pull request by silverbullet233**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Now when the query is cancelled, spill operator needs to rely on the status of spill io tasks to decide whether it can be finished, this logic is implemented through pending_finish interface which is a bit complicated.
Moreover, if the io tasks cannot be completed for a long time(e.g. under heavy load, many io tasks may be in queuing), the pipeline operator can't finish too.

In order to simplify the cancel process of spilling, the idea of scan operator is adopted here(introduced by https://github.com/StarRocks/starrocks/pull/5547). We capture the weak ptr of query context in spill tasks to determine whether the query has been canceled. If the query has been canceled, the spill task can be returned directly. In this way, the spillable operator and the spill task are decoupled, so that the operator does not need wait for the spill task to finish
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp [hunk 1]
```diff
@@ -50,9 +50,19 @@ bool SpillableAggregateBlockingSourceOperator::is_finished() const {
...
     }
+    if (_aggregator->spiller()->is_cancel()) {
+        return true;
+    }
     return _aggregator->is_spilled_eos() && !_has_last_chunk;
 }
 
+Status SpillableAggregateBlockingSourceOperator::set_finishing(RuntimeState* state) {
+    if (state->is_cancelled()) {
+        _aggregator->spiller()->cancel();
```
Comment (stdpain) [inline be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp:61]: Is it safe to call cancel here?

**Review by fzhedu — COMMENTED** [review]
please add some random test about cancel

File: be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/aggregate/spillable_aggregate_blocking_source_operator.cpp:61]: > Is it safe to call cancel here?

I think it's ok. when the cancel occurs, we don’t know which pipeline will be scheduled first, the sink or the source. The purpose of the source processing is to let the spiller perceive the cancel state as soon as possible and avoid useless work. cancel itself just sets the state and will not have other effects.

**Comment by silverbullet233** [comment]
> please add some random test about cancel

good suggestion…after introducing the fault injection mechanism, I will add some failpoints for spill to test

**Review by stdpain — DISMISSED** [review]

**Review by fzhedu — DISMISSED** [review]
LGTM

**Review by fzhedu — APPROVED** [review]
LGTM

**Review by satanson — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by silverbullet233** [comment]
@mergify backport branch-3.0

**Pull request merged**